### ISSUE TITLE
Windows 10 Docker client pull fails to download a certain blob from a private registry

### ISSUE THREAD

**Issue by 1605200517** [docker/for-win#917]
**Description**

Many images we have got share a bigger blob which is about 50 MB. Other colleagues using Linux/Mac do not have this problem.
My Windows 10 Docker client (edge) fails to download this blob, and in the registry debug level log there is no error. Here is the complete log until timeout of the client.

**Steps to reproduce the issue:**
1. Have a private registry and Windows 10 Docker client
2. Try to download images

**Describe the results you received:**
The one blob is not going to be downloaded.

```
Registry log:
12.7.2017 16:43:38time="2017-07-12T14:43:38.969932257Z" level=debug msg="authorizing request" go.version=go1.6.3 http.request.host="[[MYREPOHOST]]:8080" http.request.id=97fcb1b4-9383-46e5-8c47-495bff63b042 http.request.method=GET http.request.remoteaddr="[[IP_ADDRESS]]:48598" http.request.uri="/v2/xxx/yyy/blobs/sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.31-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(windows\\))" instance.id=ede5a3fd-e163-4cfc-9bcc-95be2eaf5c7a service=registry vars.digest="sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" vars.name="xxx/yyy" version=v2.5.1
12.7.2017 16:43:38time="2017-07-12T14:43:38.974059137Z" level=debug msg=GetBlob auth.user.name=[[REPOUSER]] go.version=go1.6.3 http.request.host="[[MYREPOHOST]]:8080" http.request.id=97fcb1b4-9383-46e5-8c47-495bff63b042 http.request.method=GET http.request.remoteaddr="[[IP_ADDRESS]]:48598" http.request.uri="/v2/xxx/yyy/blobs/sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.31-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(windows\\))" instance.id=ede5a3fd-e163-4cfc-9bcc-95be2eaf5c7a service=registry vars.digest="sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" vars.name="xxx/yyy" version=v2.5.1
12.7.2017 16:43:38time="2017-07-12T14:43:38.974266018Z" level=debug msg="filesystem.URLFor(\"/docker/registry/v2/blobs/sha256/6b/6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6/data\")" auth.user.name=[[REPOUSER]] go.version=go1.6.3 http.request.host="[[MYREPOHOST]]:8080" http.request.id=97fcb1b4-9383-46e5-8c47-495bff63b042 http.request.method=GET http.request.remoteaddr="[[IP_ADDRESS]]:48598" http.request.uri="/v2/xxx/yyy/blobs/sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.31-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(windows\\))" instance.id=ede5a3fd-e163-4cfc-9bcc-95be2eaf5c7a service=registry trace.duration=30.16µs trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).URLFor" trace.id=4f552be6-3ac6-48f2-b029-0928e159d538 trace.line=189 vars.digest="sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" vars.name="xxx/yyy" version=v2.5.1
12.7.2017 16:43:38time="2017-07-12T14:43:38.974579403Z" level=debug msg="filesystem.Reader(\"/docker/registry/v2/blobs/sha256/6b/6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6/data\", 0)" auth.user.name=[[REPOUSER]] go.version=go1.6.3 http.request.host="[[MYREPOHOST]]:8080" http.request.id=97fcb1b4-9383-46e5-8c47-495bff63b042 http.request.method=GET http.request.remoteaddr="[[IP_ADDRESS]]:48598" http.request.uri="/v2/xxx/yyy/blobs/sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.31-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(windows\\))" instance.id=ede5a3fd-e163-4cfc-9bcc-95be2eaf5c7a service=registry trace.duration=48.02µs trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Reader" trace.id=571d45ee-61c6-46c3-8167-54e62e6ed296 trace.line=107 vars.digest="sha256:6bbedd9b76a496816d86a0af731ea984f40467ef8fb23be752f801cb80436ac6" vars.name="xxx/yyy" version=v2.5.1

```

**Describe the results you expected:**
Normal pull as it should.

**Additional information you deem important (e.g. issue happens only occasionally):**
See also:
https://devops.stackexchange.com/questions/1526/strange-behaviour-of-docker-client-on-windows-timeout-for-certain-image-layers

**Output of `docker version`:**

```
PS U:\> docker version
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:30:30 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
PS U:\> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.31-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.934GiB
Name: moby
ID: PSEQ:FAAB:AX35:J2P2:2RSW:XROZ:ESL5:DFDW:2GXT:GJPT:J5ES:C2CW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 25
 Goroutines: 51
 System Time: 2017-07-12T15:23:16.1555593Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false


```
**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical 

Diagnostic ID:
99071758-71EE-4434-938A-45DEF36AD8C5/2017-07-12_17-25-33

**Comment by thaJeztah** [comment]
It may help to reduce the number of concurrent pulls; you can do so in the "advanced" daemon configuration in the configuration panel, and modify [the daemon.json that's shown there](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file), for example:

```json
{
    "debug" : true,
    "experimental": true.
    "max-concurrent-downloads": 1,
    "max-concurrent-uploads": 1
}
```

the above configuration will limit docker to 1 concurrent "download", and 1 concurrent "upload"

However, a 50MB blob is not very big, and should not be a problem to download, so I'd check if there's something in your network that makes the connection that slow

**Comment by 1605200517** [comment]
@thaJeztah thanks; unfortunately this did not work for me; I have read somewhere that on Windows changes to concurrency might have no effect however?
What I cannot understand that for other clients on same network this problem can't be reproduced.
Bad enough I can't insert debug messages to the code itself and recompile it.

**Comment by 1605200517** [comment]
fixed in Version 17.07.0-ce-rc1-win21 (12927)
Channel: edge
76a8c4b
